File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/sparse/Sddmm.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  * See the NOTICE file distributed with this work for additional
 *  * information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.api.ops.impl.sparse;

import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * Sampled Dense-Dense Matrix Multiplication (SDDMM).
 *
 * <p>For each non-zero position {@code k} at row {@code i}, column {@code j} in the sparsity
 * pattern defined by (rowPtr, colIdx), computes:
 * <pre>
 *   out[k] = sum_l  D1[i, l] * D2[j, l]
 * </pre>
 *
 * <p>C++ op name: {@code sddmm}
 * <ul>
 *   <li>Inputs:  rowPtr[rows+1 INT32], colIdx[nnz INT32], D1[rows, p], D2[cols, p]</li>
 *   <li>IArgs:   rows, cols</li>
 *   <li>Output:  values[nnz]  (same dtype as D1)</li>
 * </ul>
 *
 * <p>SDDMM is used as both a primary op and a backward primitive for
 * {@link CsrSpmv} and {@link CsrSpmm}.  First-order gradients w.r.t. D1 and D2
 * are computed by {@link SddmmBp}.
 */
public class Sddmm extends DynamicCustomOp {

    /** Number of rows in the sparse sparsity pattern. */
    private long rows;
    /** Number of columns in the sparse sparsity pattern. */
    private long cols;

    /** No-arg constructor required for op-registry reflection. */
    public Sddmm() {}

    /**
     * Eager (INDArray) constructor.
     *
     * @param rowPtr 1D [rows+1] INT32 row pointers
     * @param colIdx 1D [nnz] INT32 column indices
     * @param D1     2D [rows, p] left dense factor
     * @param D2     2D [cols, p] right dense factor
     * @param rows   number of rows
     * @param cols   number of columns
     */
    public Sddmm(INDArray rowPtr, INDArray colIdx, INDArray D1, INDArray D2,
                 long rows, long cols) {
        super(new INDArray[]{rowPtr, colIdx, D1, D2}, null);
        this.rows = rows;
        this.cols = cols;
        addIArgument(rows, cols);
    }

    /**
     * SameDiff (symbolic) constructor.
     *
     * @param sd     the SameDiff graph
     * @param rowPtr SDVariable [rows+1] INT32 row pointers
     * @param colIdx SDVariable [nnz] INT32 column indices
     * @param D1     SDVariable [rows, p] left dense factor
     * @param D2     SDVariable [cols, p] right dense factor
     * @param rows   number of rows
     * @param cols   number of columns
     */
    public Sddmm(SameDiff sd, SDVariable rowPtr, SDVariable colIdx,
                 SDVariable D1, SDVariable D2, long rows, long cols) {
        super(sd, new SDVariable[]{rowPtr, colIdx, D1, D2});
        this.rows = rows;
        this.cols = cols;
        addIArgument(rows, cols);
    }

    @Override
    public String opName() { return "sddmm"; }

    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> dataTypes) {
        // Output dtype matches D1 (input[2]) dtype
        return Collections.singletonList(dataTypes.get(2));
    }

    /**
     * Backward pass for sddmm — delegates to the dedicated {@code sddmm_bp} op.
     *
     * <p>Outputs of the backward op:
     * <ul>
     *   <li>[0] dD1[rows, p]: gradient w.r.t. the left dense factor D1</li>
     *   <li>[1] dD2[cols, p]: gradient w.r.t. the right dense factor D2</li>
     * </ul>
     * rowPtr and colIdx are integer/structural and receive zero gradients.
     */
    @Override
    public List<SDVariable> doDiff(List<SDVariable> grads) {
        SDVariable gradOut = grads.get(0);

        long r = iArguments.size() > 0 ? iArguments.get(0) : this.rows;
        long c = iArguments.size() > 1 ? iArguments.get(1) : this.cols;

        SDVariable rowPtrVar = arg(0);
        SDVariable colIdxVar = arg(1);
        SDVariable D1Var     = arg(2);
        SDVariable D2Var     = arg(3);

        // Delegate to dedicated backward op: outputs are [dD1, dD2]
        SDVariable[] bpOut = new SddmmBp(sameDiff, rowPtrVar, colIdxVar, D1Var, D2Var, gradOut, r, c)
                .outputVariables();

        return Arrays.asList(sameDiff.zerosLike(rowPtrVar), sameDiff.zerosLike(colIdxVar),
                bpOut[0], bpOut[1]);
    }
}

File: libnd4j/include/ops/declarable/generic/sparse/sddmm.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// SDDMM — sampled dense-dense matrix multiplication.
// Computes output values at the CSR sparsity pattern:
//
//   outValues[k] = sum_{l=0}^{p-1} D1[i,l] * D2[j,l]
//
// where i is the row of nonzero k (rowPtr[i] <= k < rowPtr[i+1])
// and j = colIdx[k].
// Equivalently: outValues = (D1 · D2^T) sampled at the (rowPtr, colIdx) pattern.
//
// Inputs:
//   [0] rowPtr    – 1D [rows+1], INT32/INT64  (row pointer array)
//   [1] colIdx    – 1D [nnz],   same INT       (column indices)
//   [2] D1        – dense 2D [rows, p]
//   [3] D2        – dense 2D [cols, p]
// IArgs:
//   [0] rows
//   [1] cols
// Output:
//   [0] outValues – 1D [nnz], same float dtype as D1
//

#include <system/op_boilerplate.h>
#if NOT_EXCLUDED(OP_sddmm)

#include <ops/declarable/headers/sparse.h>
#include <ops/declarable/helpers/sparse_blas.h>

namespace sd {
namespace ops {

CUSTOM_OP_IMPL(sddmm, 4, 1, false, 0, 2) {
  auto rowPtr    = INPUT_VARIABLE(0);
  auto colIdx    = INPUT_VARIABLE(1);
  auto D1        = INPUT_VARIABLE(2);
  auto D2        = INPUT_VARIABLE(3);

  const LongType rows = INT_ARG(0);
  const LongType cols = INT_ARG(1);

  REQUIRE_TRUE(rowPtr->rankOf() == 1, 0,
               "sddmm: rowPtr must be 1D, got rank %d", rowPtr->rankOf());
  REQUIRE_TRUE(colIdx->rankOf() == 1, 0,
               "sddmm: colIdx must be 1D, got rank %d", colIdx->rankOf());
  REQUIRE_TRUE(D1->rankOf() == 2, 0,
               "sddmm: D1 must be 2D [rows, p], got rank %d", D1->rankOf());
  REQUIRE_TRUE(D2->rankOf() == 2, 0,
               "sddmm: D2 must be 2D [cols, p], got rank %d", D2->rankOf());
  REQUIRE_TRUE(rowPtr->lengthOf() == rows + 1, 0,
               "sddmm: rowPtr length must equal rows+1=%lld, got %lld",
               (long long)(rows + 1), (long long)rowPtr->lengthOf());
  REQUIRE_TRUE(rowPtr->dataType() == colIdx->dataType(), 0,
               "sddmm: rowPtr and colIdx must have the same integer dtype");
  REQUIRE_TRUE(D1->sizeAt(0) == rows, 0,
               "sddmm: D1 must have %lld rows, got %lld",
               (long long)rows, (long long)D1->sizeAt(0));
  REQUIRE_TRUE(D2->sizeAt(0) == cols, 0,
               "sddmm: D2 must have %lld rows (= cols of A), got %lld",
               (long long)cols, (long long)D2->sizeAt(0));
  REQUIRE_TRUE(D1->sizeAt(1) == D2->sizeAt(1), 0,
               "sddmm: D1 and D2 must share the inner dimension p: %lld vs %lld",
               (long long)D1->sizeAt(1), (long long)D2->sizeAt(1));

  auto outValues = OUTPUT_NULLIFIED(0);

  sd::ops::helpers::sddmm(*rowPtr, *colIdx, *D1, *D2, *outValues, rows, cols);

  return sd::Status::OK;
}

DECLARE_SHAPE_FN(sddmm) {
  auto colIdx = INPUT_VARIABLE(1);

  // nnz is the length of colIdx
  const LongType nnz = colIdx->lengthOf();

  auto D1 = INPUT_VARIABLE(2);
  auto dtype    = D1->dataType();
  auto outShape = ConstantShapeHelper::getInstance().createShapeInfo(dtype, 'c', {nnz});
  return SHAPELIST(outShape);
}

DECLARE_TYPES(sddmm) {
  getOpDescriptor()
      ->setAllowedInputTypes(0, {ALL_INTS})         // rowPtr
      ->setAllowedInputTypes(1, {ALL_INTS})         // colIdx
      ->setAllowedInputTypes(2, {ALL_FLOATS})   // D1
      ->setAllowedInputTypes(3, {ALL_FLOATS})   // D2
      ->setAllowedOutputTypes({ALL_FLOATS});
}

}  // namespace ops
}  // namespace sd

#endif  // NOT_EXCLUDED(OP_sddmm)
